File: kavach-node/npm/src/http.ts
/**
 * Kavach middleware for Node.js HTTP frameworks.
 *
 * Works with Express, Fastify, Hono, and any framework
 * that supports middleware functions.
 *
 * @example Express
 * ```typescript
 * import express from 'express';
 * import { Gate, createExpressMiddleware } from 'kavach';
 *
 * const gate = Gate.fromFile('kavach.toml');
 * const app = express();
 *
 * // Gate all mutating requests
 * app.use(createExpressMiddleware(gate));
 *
 * app.post('/api/refunds', (req, res) => {
 *   // If we reach here, Kavach permitted the action
 *   res.json({ status: 'refunded' });
 * });
 * ```
 */

import {
  Gate,
  KavachRefused,
  KavachInvalidated,
  type PrincipalKind,
  type GeoLocationInput,
} from './index';
import type { VerdictResult } from 'kavach-engine';

export interface HttpMiddlewareOptions {
  /** Only gate mutating requests (POST/PUT/DELETE/PATCH). Default: true */
  gateMutationsOnly?: boolean;
  /** Paths to exclude from gating. Default: ['/health', '/ready'] */
  excludedPaths?: string[];
  /** Header containing principal ID. Default: 'x-principal-id' */
  principalHeader?: string;
  /** Header containing roles (comma-separated). Default: 'x-roles' */
  rolesHeader?: string;
  /** Header containing principal kind. Default: 'x-principal-kind' */
  kindHeader?: string;
  /**
   * Optional geo resolver — integrators plug in their GeoIP lookup
   * (MaxMind, CDN edge headers, etc.) to populate `currentGeo` and
   * `originGeo`. Return `undefined` for either to leave it unset.
   * Needed to drive tolerant-mode `GeoLocationDrift`.
   */
  geoResolver?: (req: {
    method: string;
    path: string;
    headers: Record<string, string | string[] | undefined>;
    ip?: string;
  }) => { currentGeo?: GeoLocationInput; originGeo?: GeoLocationInput } | undefined;
}

const VERB_MAP: Record<string, string> = {
  GET: 'read',
  POST: 'create',
  PUT: 'update',
  PATCH: 'update',
  DELETE: 'delete',
};

/**
 * Derive an action name from HTTP method + path.
 * POST /api/v1/refunds → refunds.create
 */
export function deriveActionName(method: string, path: string): string {
  const parts = path
    .split('/')
    .filter(p => p && p !== 'api' && !p.startsWith('v') && !/^\d+$/.test(p));
  const resource = parts[parts.length - 1] || 'unknown';
  const verb = VERB_MAP[method.toUpperCase()] || 'unknown';
  return `${resource}.${verb}`;
}

/**
 * Core HTTP gate — framework agnostic.
 */
export class HttpKavachMiddleware {
  private gate: Gate;
  private opts: Required<HttpMiddlewareOptions>;

  constructor(gate: Gate, options: HttpMiddlewareOptions = {}) {
    this.gate = gate;
    this.opts = {
      gateMutationsOnly: options.gateMutationsOnly ?? true,
      excludedPaths: options.excludedPaths ?? ['/health', '/ready', '/metrics'],
      principalHeader: options.principalHeader ?? 'x-principal-id',
      rolesHeader: options.rolesHeader ?? 'x-roles',
      kindHeader: options.kindHeader ?? 'x-principal-kind',
      geoResolver: options.geoResolver ?? (() => undefined),
    };
  }

  shouldGate(method: string, path: string): boolean {
    if (this.opts.excludedPaths.some(p => path.startsWith(p))) return false;
    if (this.opts.gateMutationsOnly && method.toUpperCase() === 'GET') return false;
    return true;
  }

  evaluate(req: {
    method: string;
    path: string;
    headers: Record<string, string | string[] | undefined>;
    body?: Record<string, unknown>;
    ip?: string;
    /**
     * Current geographic location for this request (→ `EnvContext.geo`).
     * Integrators populate from GeoIP / CDN headers / edge metadata.
     */
    currentGeo?: GeoLocationInput;
    /**
     * Geographic location captured at session start
     * (→ `SessionState.origin_geo`). Integrators pass this from
     * wherever session state lives.
     */
    originGeo?: GeoLocationInput;
  }): VerdictResult {
    const header = (name: string) => {
      const v = req.headers[name];
      return typeof v === 'string' ? v : v?.[0] ?? '';
    };

    const numericParams: Record<string, number> = {};
    if (req.body) {
      for (const [k, v] of Object.entries(req.body)) {
        if (typeof v === 'number') numericParams[k] = v;
      }
    }

    // If the caller didn't pass explicit geo, fall back to the configured
    // resolver. Keeps factories and direct callers symmetric.
    const resolved =
      req.currentGeo === undefined && req.originGeo === undefined
        ? this.opts.geoResolver({
            method: req.method,
            path: req.path,
            headers: req.headers,
            ip: req.ip,
          })
        : undefined;

    return this.gate.evaluate({
      principalId: header(this.opts.principalHeader) || 'anonymous',
      principalKind: (header(this.opts.kindHeader) || 'user') as PrincipalKind,
      actionName: deriveActionName(req.method, req.path),
      roles: header(this.opts.rolesHeader).split(',').filter(Boolean),
      resource: req.path,
      params: Object.keys(numericParams).length > 0 ? numericParams : undefined,
      ip: req.ip,
      currentGeo: req.currentGeo ?? resolved?.currentGeo,
      originGeo: req.originGeo ?? resolved?.originGeo,
    });
  }
}

// ─── Express middleware factory ──────────────────────────────────

/**
 * Create Express middleware that gates requests through Kavach.
 *
 * @example
 * ```typescript
 * app.use(createExpressMiddleware(gate, { gateMutationsOnly: true }));
 * ```
 */
export function createExpressMiddleware(
  gate: Gate,
  options?: HttpMiddlewareOptions,
) {
  const middleware = new HttpKavachMiddleware(gate, options);

  return (req: any, res: any, next: any) => {
    if (!middleware.shouldGate(req.method, req.path)) {
      return next();
    }

    const verdict = middleware.evaluate({
      method: req.method,
      path: req.path,
      headers: req.headers,
      body: req.body,
      ip: req.ip,
    });

    if (verdict.isPermit) {
      return next();
    }

    if (verdict.isInvalidate) {
      return res.status(401).json({
        error: 'session_invalidated',
        reason: verdict.reason,
      });
    }

    return res.status(403).json({
      error: 'forbidden',
      code: verdict.code,
      reason: verdict.reason,
    });
  };
}

// ─── Fastify plugin factory ─────────────────────────────────────

/**
 * Create a Fastify onRequest hook that gates requests through Kavach.
 *
 * @example
 * ```typescript
 * fastify.addHook('onRequest', createFastifyHook(gate));
 * ```
 */
export function createFastifyHook(
  gate: Gate,
  options?: HttpMiddlewareOptions,
) {
  const middleware = new HttpKavachMiddleware(gate, options);

  return (request: any, reply: any, done: any) => {
    if (!middleware.shouldGate(request.method, request.url)) {
      return done();
    }

    const verdict = middleware.evaluate({
      method: request.method,
      path: request.url,
      headers: request.headers,
      body: request.body,
      ip: request.ip,
    });

    if (verdict.isPermit) {
      return done();
    }

    reply.code(verdict.isInvalidate ? 401 : 403).send({
      error: verdict.isInvalidate ? 'session_invalidated' : 'forbidden',
      code: verdict.code,
      reason: verdict.reason,
    });
  };
}

File: kavach-py/python/kavach/http.py
"""Kavach middleware for Python HTTP frameworks.

Works with FastAPI, Starlette, Flask, and any ASGI/WSGI framework.
All evaluation runs in the compiled Rust engine.

Example with FastAPI:
    from fastapi import FastAPI, Request, HTTPException
    from kavach import Gate, HttpKavachMiddleware

    gate = Gate.from_file("kavach.toml")
    kavach = HttpKavachMiddleware(gate)

    app = FastAPI()

    @app.middleware("http")
    async def kavach_middleware(request: Request, call_next):
        verdict = kavach.evaluate_request(request)
        if not verdict.is_permit:
            raise HTTPException(status_code=403, detail=verdict.reason)
        return await call_next(request)
"""

from typing import Callable, TypedDict

from kavach._kavach_engine import ActionContext, GeoLocation, Verdict
from kavach.wrappers import Gate


class GeoResolverResult(TypedDict, total=False):
    """Optional geo annotations returned by a ``geo_resolver``."""

    current_geo: GeoLocation | None
    origin_geo: GeoLocation | None


class HttpKavachMiddleware:
    """Kavach middleware for HTTP APIs.

    Translates HTTP requests to ActionContext and evaluates
    through the Rust gate. Framework-specific adapters call
    into this class.
    """

    def __init__(
        self,
        gate: Gate,
        *,
        gate_mutations_only: bool = True,
        excluded_paths: list[str] | None = None,
        principal_header: str = "X-Principal-Id",
        roles_header: str = "X-Roles",
        kind_header: str = "X-Principal-Kind",
        geo_resolver: Callable[..., GeoResolverResult | None] | None = None,
    ):
        """Args:
            geo_resolver: Optional callable invoked as
                ``geo_resolver(method=..., path=..., ip=..., headers=...)``
                returning a dict with ``current_geo`` / ``origin_geo`` keys
                (both optional ``GeoLocation``). Integrators plug in their
                GeoIP lookup here to drive tolerant-mode
                ``GeoLocationDrift``.
        """
        self._gate = gate
        self.gate_mutations_only = gate_mutations_only
        self.excluded_paths = excluded_paths or ["/health", "/ready", "/metrics"]
        self.principal_header = principal_header
        self.roles_header = roles_header
        self.kind_header = kind_header
        self._geo_resolver = geo_resolver

    def should_gate(self, method: str, path: str) -> bool:
        """Check if this request should be gated."""
        if any(path.startswith(p) for p in self.excluded_paths):
            return False
        if self.gate_mutations_only and method.upper() == "GET":
            return False
        return True

    def derive_action_name(self, method: str, path: str) -> str:
        """Derive action name from HTTP method + path.

        POST /api/v1/refunds → refunds.create
        DELETE /api/v1/users/123 → users.delete
        """
        parts = [
            p for p in path.split("/")
            if p and p != "api" and not p.startswith("v")
            and not p.isdigit()
        ]
        resource = parts[-1] if parts else "unknown"

        verb_map = {
            "GET": "read", "POST": "create",
            "PUT": "update", "PATCH": "update",
            "DELETE": "delete",
        }
        verb = verb_map.get(method.upper(), "unknown")
        return f"{resource}.{verb}"

    def evaluate(
        self,
        *,
        method: str,
        path: str,
        principal_id: str = "anonymous",
        principal_kind: str = "user",
        roles: list[str] | None = None,
        body: dict | None = None,
        ip: str | None = None,
        session_id: str | None = None,
        current_geo: GeoLocation | None = None,
        origin_geo: GeoLocation | None = None,
    ) -> Verdict:
        """Evaluate an HTTP request against the gate.

        Returns the Verdict, caller decides how to handle it. Pass
        ``current_geo`` / ``origin_geo`` (a ``GeoLocation`` with at
        least ``country_code`` and, for tolerant-mode drift,
        ``latitude``/``longitude``) to feed ``GeoLocationDrift``.
        """
        if not self.should_gate(method, path):
            # Return a synthetic permit for non-gated requests
            from kavach._kavach_engine import Verdict as _V
            # We can't construct a Verdict from Python, so we
            # just evaluate with a permissive context
            pass

        action_name = self.derive_action_name(method, path)

        numeric_params = {}
        if body:
            numeric_params = {
                k: float(v) for k, v in body.items()
                if isinstance(v, (int, float))
            }

        # Fall back to the configured geo_resolver if no explicit geo.
        if (
            current_geo is None
            and origin_geo is None
            and self._geo_resolver is not None
        ):
            resolved = self._geo_resolver(method=method, path=path, ip=ip) or {}
            current_geo = resolved.get("current_geo")
            origin_geo = resolved.get("origin_geo")

        ctx = ActionContext(
            principal_id=principal_id,
            principal_kind=principal_kind,
            action_name=action_name,
            roles=roles or [],
            resource=path,
            params=numeric_params or None,
            ip=ip,
            session_id=session_id,
            current_geo=current_geo,
            origin_geo=origin_geo,
        )

        return self._gate.evaluate(ctx)

    def check(self, **kwargs) -> None:
        """Evaluate and raise if not permitted.

        Same args as evaluate(). Raises Refused or Invalidated.
        """
        verdict = self.evaluate(**kwargs)
        if verdict.is_refuse:
            from kavach.wrappers import Refused
            raise Refused(verdict.reason, verdict.evaluator, verdict.code)
        if verdict.is_invalidate:
            from kavach.wrappers import Invalidated
            raise Invalidated(verdict.reason, verdict.evaluator)

    # ── Framework-specific helpers ────────────────────────────────

    def evaluate_fastapi(self, request) -> Verdict:
        """Evaluate a FastAPI/Starlette Request object.

        Usage:
            @app.middleware("http")
            async def kavach_gate(request: Request, call_next):
                verdict = kavach.evaluate_fastapi(request)
                if not verdict.is_permit:
                    return JSONResponse(status_code=403, content={"error": verdict.reason})
                return await call_next(request)
        """
        headers = dict(request.headers)
        return self.evaluate(
            method=request.method,
            path=request.url.path,
            principal_id=headers.get(self.principal_header.lower(), "anonymous"),
            principal_kind=headers.get(self.kind_header.lower(), "user"),
            roles=(headers.get(self.roles_header.lower(), "")).split(","),
            ip=request.client.host if request.client else None,
        )

    def evaluate_flask(self, request) -> Verdict:
        """Evaluate a Flask request object.

        Usage:
            @app.before_request
            def kavach_gate():
                verdict = kavach.evaluate_flask(request)
                if not verdict.is_permit:
                    return jsonify(error=verdict.reason), 403
        """
        return self.evaluate(
            method=request.method,
            path=request.path,
            principal_id=request.headers.get(self.principal_header, "anonymous"),
            principal_kind=request.headers.get(self.kind_header, "user"),
            roles=request.headers.get(self.roles_header, "").split(","),
            body=request.get_json(silent=True),
            ip=request.remote_addr,
        )
